rskj-core/src/main/java/co/rsk/net/AsyncNodeBlockProcessor.java

```diff
@@ -0,0 +1,227 @@
+
+    private final BlockingQueue<BlockInfo> blocksToProcess = new LinkedBlockingQueue<>();
+
+    private final BlockValidator blockHeaderValidator;
+
+    private final BlockValidator blockValidator;
+
+    private final Thread thread = new Thread(this,"async block processor");
+
+    private final Listener listener;
+
+    private volatile boolean stopped;
+
+    public AsyncNodeBlockProcessor(@Nonnull NetBlockStore store, @Nonnull Blockchain blockchain, @Nonnull BlockNodeInformation nodeInformation,
+                                   @Nonnull BlockSyncService blockSyncService, @Nonnull SyncConfiguration syncConfiguration,
+                                   @Nonnull BlockValidator blockHeaderValidator, @Nonnull BlockValidator blockValidator,
+                                   @Nullable Listener listener) {
+        super(store, blockchain, nodeInformation, blockSyncService, syncConfiguration);
+        this.blockHeaderValidator = blockHeaderValidator;
+        this.blockValidator = blockValidator;
+        this.listener = listener;
+    }
+
+    public AsyncNodeBlockProcessor(@Nonnull NetBlockStore store, @Nonnull Blockchain blockchain, @Nonnull BlockNodeInformation nodeInformation,
+                                   @Nonnull BlockSyncService blockSyncService, @Nonnull SyncConfiguration syncConfiguration,
+                                   @Nonnull BlockValidator blockHeaderValidator, @Nonnull BlockValidator blockValidator) {
+        this(store, blockchain, nodeInformation, blockSyncService, syncConfiguration, blockHeaderValidator, blockValidator, null);
+    }
+
+    @Override
+    public BlockProcessResult processBlock(@Nullable Peer sender, @Nonnull Block block) {
+        final Instant start = Instant.now();
+
+        final long blockNumber = block.getNumber();
+        final String blockHash = block.getPrintableHash();
+        final String peer = sender != null ? sender.getPeerNodeID().toString() : "N/A";
+
+        // Validate block header first to see if its PoW is valid at all
+        if (!isBlockHeaderValid(block)) {
+            logger.warn("Invalid block with number {} {} from {} ", blockNumber, blockHash, peer);
+            return invalidBlockResult(block, start);
+        }
+
+        // Check if block is already in the queue
+        if (store.hasBlock(block)) {
+            logger.trace("Ignored block with number {} and hash {} from {} as it's already in the queue", blockNumber, blockHash, peer);
+            return ignoreBlockResult(block, start);
+        }
+
+        // Check if block is ready for processing - if the block is not too advanced, its parent block is in place etc.
+        boolean readyForProcessing = blockSyncService.preprocessBlock(block, sender, false);
+        if (readyForProcessing) {
+            // Validate block if it can be added to the queue for processing
+            if (isBlockValid(block)) {
+                boolean offer = blocksToProcess.offer(new BlockInfo(sender, block));
```

Comment:
I would separate the logic of "adding to the queue" in another method, and check for "stopped" flag.
Then you could decide to log(that it is stopped) or log/stop the node(?)